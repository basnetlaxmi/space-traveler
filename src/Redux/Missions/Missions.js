import { fetchMissionsData } from '../apiFunctions';
// Constant
const FETCH_MISSIONS = 'space-traveler/Missions/FETCH_MISSIONS';
const JOIN_MISSION = 'space-traveler/Missions/JOIN_MISSION';
const LEAVE_MISSION = 'space-traveler/Missions/LEAVE_MISSION';
// Action
export const fetchMission = () => async (dispatch) => {
  const res = await fetchMissionsData();
  dispatch({ type: FETCH_MISSIONS, payload: res });
};

export const joinMission = (id) => ({
  type: JOIN_MISSION,
  payload: id,
});

export const leaveMission = (id) => ({
  type: LEAVE_MISSION,
  payload: id,
});

// Reducers
const initialState = {
  missions: [],
  joined: [],

};
const missionReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_MISSIONS:
      return {
        ...state,
        missions: action.payload,

      };
    case JOIN_MISSION:
    {
      const newState = state.missions.map((mission) => {
        if (mission.id !== action.payload) {
          return mission;
        }
        return { ...mission, reserved: true };
      });
      return {
        ...state,
        missions: [...newState],
      };
    }
    case LEAVE_MISSION:
    {
      const newState = state.missions.map((mission) => {
        if (mission.id !== action.payload) {
          return mission;
        }
        return { ...mission, reserved: false };
      });
      return {
        ...state,
        missions: [...newState],
      };
    }
    default:
      return state;
  }
};

export default missionReducer;
