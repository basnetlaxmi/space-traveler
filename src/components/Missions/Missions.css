body {
  overflow-x: hidden;
}

.table {
    --bs-table-bg: transparent;
    --bs-table-accent-bg: transparent;
    --bs-table-striped-color: #212529;
    --bs-table-striped-bg: rgba(0,0,0,.05);
    --bs-table-active-color: #212529;
    --bs-table-active-bg: rgba(0,0,0,.1);
    --bs-table-hover-color: #212529;
    --bs-table-hover-bg: rgba(0,0,0,.075);
    border-color: #dee2e6;
    color: #212529;
    margin-bottom: 1rem;
    vertical-align: top;
    width: 100%;
}
table {
    border-collapse: collapse;
    caption-side: bottom;
}
.table>thead {
    vertical-align: bottom;
}
tbody, td, tfoot, th, thead, tr {
    border: 0 solid;
    border-color: inherit;
}
.table>:not(:first-child) {
    border-top: 2px solid;
}
.table>tbody {
    vertical-align: inherit;
}
.table-bordered>:not(caption)>* {
    border-width: 1px 0;
}
.table-striped>tbody>tr:nth-of-type(odd)>* {
    --bs-table-accent-bg: var(--bs-table-striped-bg);
    color: var(--bs-table-striped-color);
}
.table-bordered>:not(caption)>*>* {
    border-width: 0 1px;
}
.table>:not(caption)>*>* {
    background-color: var(--bs-table-bg);
    border-bottom-width: 1px;
    box-shadow: inset 0 0 0 9999px var(--bs-table-accent-bg);
    padding: 0.5rem;
}
.member-btn {
  background-color: #828990;
  color: #fff;
  width: 150px;
  padding: 0;
  margin: 50px auto;
  border-radius: 5px;
}

.join-leave-btn {
  width: 150px;
  padding: 5px;
  margin: 50px auto;
  border-radius: 5px;
}
.leave-btn {
  color: #de4956;
  border-color: #de4956;
}

.active-member {
  background-color: #18a2b8;
  color: #fff;
  border-color: #18a2b8;
}
