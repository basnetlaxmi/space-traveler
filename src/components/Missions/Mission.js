import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { joinMission, leaveMission } from '../../Redux/Missions/Missions';

const Mission = ({ mission }) => {
  const dispatch = useDispatch();
  return (
    <div className="flex flex-column pb4 items-start-ns items-center" style={{ gap: '0.75rem', borderBottom: '1px solid grey' }}>
      <p className="ma0"><b>{mission.name}</b></p>
      <p className="mb1 mt1 tl-ns tc">{mission.description}</p>
      {mission.reserved ? <button type="button" className="member-btn text-center active-member">Active Member</button>
        : <button type="button" className="member-btn text-center">Not a member</button>}
      {mission.reserved
        ? <button type="button" className="join-leave-btn pointer leave-btn" onClick={() => dispatch(leaveMission(mission.id))}>Leave Mission</button>
        : <button type="button" className="join-leave-btn pointer" onClick={() => dispatch(joinMission(mission.id))}>Join Mission</button>}
    </div>
  );
};
Mission.propTypes = {
  mission: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  ).isRequired,
};
export default Mission;
