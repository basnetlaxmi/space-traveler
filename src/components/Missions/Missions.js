import { useSelector } from 'react-redux';
import './Missions.css';
import Mission from './Mission';

const Missions = () => {
  const missions = useSelector((state) => state.missionReducer.missions);
  return (
    <section>
      <p>
        <span>Mission</span>
        <span>Description</span>
        <span>Status</span>
        <span>Action</span>
      </p>
      {missions.map((mission) => <Mission mission={mission} key={mission.id} />)}
    </section>
  );
};

export default Missions;
