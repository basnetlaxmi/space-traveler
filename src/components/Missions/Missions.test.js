import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../Redux/createStore';
import Missions from './Missions';
import Mission from './Mission';

describe('Missions', () => {
  const mission = {
    id: '12345',
    name: 'Telstar',
    description: 'This is a description of Telstar',
    reserved: false,
  };
  render(
    <Provider store={store}>
      <Mission mission={mission} />
    </Provider>,
  );
  const renderMissions = render(<Provider store={store}><Missions /></Provider>);
  test('Take snapshot for Missions', () => {
    expect(renderMissions).toMatchSnapshot();
    const table = screen.getAllByRole('row');
    expect(table[0]).toBeInTheDocument();
  });
});
