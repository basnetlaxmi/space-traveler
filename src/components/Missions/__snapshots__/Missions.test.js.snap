// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Take snapshot for Missions 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
       
      <tr>
        <td>
          <b>
            Telstar
          </b>
        </td>
        <td>
          This is a description of Telstar
        </td>
        <td>
          <button
            class="member-btn text-center"
            type="button"
          >
            Not a member
          </button>
        </td>
        <td>
          <button
            class="join-leave-btn"
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
    </div>
    <div>
      <div
        class="m-4"
      >
        <table
          class="table table-striped table-bordered"
          role="row"
        >
          <thead>
            <tr>
              <th>
                Mission
              </th>
              <th>
                Description
              </th>
              <th>
                Status
              </th>
              <th />
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="m-4"
    >
      <table
        class="table table-striped table-bordered"
        role="row"
      >
        <thead>
          <tr>
            <th>
              Mission
            </th>
            <th>
              Description
            </th>
            <th>
              Status
            </th>
            <th />
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
